<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Bibliography &mdash; MetaGenePipe 1.0.0 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="_static/copybutton.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/clipboard.min.js"></script>
        <script src="_static/copybutton.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Contributing" href="contributing.html" />
    <link rel="prev" title="Citation and Attribution" href="citation.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="index.html" class="icon icon-home"> MetaGenePipe
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="usage.html">Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="workflow.html">Workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="extension.html">Extension</a></li>
<li class="toctree-l1"><a class="reference internal" href="citation.html">Citation and Attribution</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Bibliography</a></li>
<li class="toctree-l1"><a class="reference internal" href="contributing.html">Contributing</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">MetaGenePipe</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
      <li>Bibliography</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/bibliography.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="bibliography">
<h1>Bibliography<a class="headerlink" href="#bibliography" title="Permalink to this headline"></a></h1>
<p>The following articles and software packages were used in the development of MetaGenePipe.</p>
<div class="docutils container" id="id1">
<dl class="citation">
<dt class="label" id="id25"><span class="brackets">kra</span></dt>
<dd><p>Kraken taxonomic sequence classification system: operating manual. URL: <a class="reference external" href="https://ccb.jhu.edu/software/kraken/MANUAL.html">https://ccb.jhu.edu/software/kraken/MANUAL.html</a>, <a class="reference external" href="https://doi.org/10.1186/gb-2014-15-3-r46">doi:10.1186/gb-2014-15-3-r46</a>.</p>
</dd>
<dt class="label" id="id16"><span class="brackets">AGM+90</span></dt>
<dd><p>S F Altschul, W Gish, W Miller, E W Myers, and D J Lipman. Basic local alignment search tool. <em>J. Mol. Biol.</em>, 215(3):403–410, October 1990. <a class="reference external" href="https://doi.org/10.1016/S0022-2836(05)80360-2">doi:10.1016/S0022-2836(05)80360-2</a>.</p>
</dd>
<dt class="label" id="id12"><span class="brackets">AMSchaffer+97</span></dt>
<dd><p>S F Altschul, T L Madden, A A Schäffer, J Zhang, Z Zhang, W Miller, and D J Lipman. Gapped BLAST and PSI-BLAST: a new generation of protein database search programs. <em>Nucleic Acids Research</em>, 25(17):3389–3402, September 1997. <a class="reference external" href="https://doi.org/10.1093/nar/25.17.3389">doi:10.1093/nar/25.17.3389</a>.</p>
</dd>
<dt class="label" id="id5"><span class="brackets">And10</span></dt>
<dd><p>S. Andrews. FASTQC. A quality control tool for high throughput sequence data. 2010. URL: <a class="reference external" href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">https://www.bioinformatics.babraham.ac.uk/projects/fastqc/</a>.</p>
</dd>
<dt class="label" id="id17"><span class="brackets">ABME+20</span></dt>
<dd><p>T. Aramaki, R. Blanc-Mathieu, H. Endo, K. Ohkubo, M. Kanehisa, S. Goto, and H. Ogata. KofamKOALA: KEGG Ortholog assignment based on profile HMM and adaptive score threshold. <em>Bioinformatics</em>, 36(7):2251–2252, 04 2020. <a class="reference external" href="https://doi.org/10.1093/bioinformatics/btz859">doi:10.1093/bioinformatics/btz859</a>.</p>
</dd>
<dt class="label" id="id42"><span class="brackets">ABME+19</span></dt>
<dd><p>Takuya Aramaki, Romain Blanc-Mathieu, Hisashi Endo, Koichi Ohkubo, Minoru Kanehisa, Susumu Goto, and Hiroyuki Ogata. Kofamkoala: kegg ortholog assignment based on profile hmm and adaptive score threshold. <em>Bioinformatics</em>, 36(7):2251–2252, 2019. <a class="reference external" href="https://doi.org/10.1093/bioinformatics/btz859">doi:10.1093/bioinformatics/btz859</a>.</p>
</dd>
<dt class="label" id="id2"><span class="brackets">AHS+17</span></dt>
<dd><p>Katherine E Arden, Claire Heney, Babak Shaban, Graeme R Nimmo, Michael D Nissen, Theo P Sloots, and Ian M Mackay. Detection of Toscana virus from an adult traveler returning to Australia with encephalitis. <em>J. Med. Virol.</em>, 89(10):1861–1864, October 2017. <a class="reference external" href="https://doi.org/10.1002/jmv.24839">doi:10.1002/jmv.24839</a>.</p>
</dd>
<dt class="label" id="id29"><span class="brackets">BGQ+11</span></dt>
<dd><p>Derek W. Barnett, Erik K. Garrison, Aaron R. Quinlan, Michael P. Strömberg, and Gabor T. Marth. BamTools: a C++ API and toolkit for analyzing and managing BAM files. <em>Bioinformatics</em>, 27(12):1691–1692, 04 2011. URL: <a class="reference external" href="https://doi.org/10.1093/bioinformatics/btr174">https://doi.org/10.1093/bioinformatics/btr174</a>, <a class="reference external" href="https://arxiv.org/abs/https://academic.oup.com/bioinformatics/article-pdf/27/12/1691/709404/btr174.pdf">arXiv:https://academic.oup.com/bioinformatics/article-pdf/27/12/1691/709404/btr174.pdf</a>, <a class="reference external" href="https://doi.org/10.1093/bioinformatics/btr174">doi:10.1093/bioinformatics/btr174</a>.</p>
</dd>
<dt class="label" id="id27"><span class="brackets">BLU14</span></dt>
<dd><p>A. M. Bolger, M. Lohse, and B. Usadel. Trimmomatic: a flexible trimmer for Illumina sequence data. <em>Bioinformatics</em>, 30(15):2114–2120, Aug 2014. <a class="reference external" href="https://doi.org/10.1093/bioinformatics/btu170">doi:10.1093/bioinformatics/btu170</a>.</p>
</dd>
<dt class="label" id="id28"><span class="brackets">BLT+07</span></dt>
<dd><p>E. Boutet, D. Lieberherr, M. Tognolli, M. Schneider, and A. Bairoch. UniProtKB/Swiss-Prot. <em>Methods Mol Biol</em>, 406:89–112, 2007. <a class="reference external" href="https://doi.org/10.1007/978-1-59745-535-0_4">doi:10.1007/978-1-59745-535-0_4</a>.</p>
</dd>
<dt class="label" id="id18"><span class="brackets">BLB20</span></dt>
<dd><p>Tomáš Brůna, Alexandre Lomsadze, and Mark Borodovsky. GeneMark-EP+: eukaryotic gene prediction with self-training in the space of genes and proteins. <em>NAR Genomics and Bioinformatics</em>, 05 2020. lqaa026. URL: <a class="reference external" href="https://doi.org/10.1093/nargab/lqaa026">https://doi.org/10.1093/nargab/lqaa026</a>, <a class="reference external" href="https://arxiv.org/abs/https://academic.oup.com/nargab/article-pdf/2/2/lqaa026/34054524/lqaa026.pdf">arXiv:https://academic.oup.com/nargab/article-pdf/2/2/lqaa026/34054524/lqaa026.pdf</a>, <a class="reference external" href="https://doi.org/10.1093/nargab/lqaa026">doi:10.1093/nargab/lqaa026</a>.</p>
</dd>
<dt class="label" id="id44"><span class="brackets">BRD21</span></dt>
<dd><p>Benjamin Buchfink, Klaus Reuter, and Hajk-Georg Drost. Sensitive protein alignments at tree-of-life scale using diamond. <em>Nature Methods</em>, 18(4):366–368, 2021. <a class="reference external" href="https://doi.org/10.1038/s41592-021-01101-x">doi:10.1038/s41592-021-01101-x</a>.</p>
</dd>
<dt class="label" id="id9"><span class="brackets">BXH15</span></dt>
<dd><p>Benjamin Buchfink, Chao Xie, and Daniel H Huson. Fast and sensitive protein alignment using DIAMOND. <em>Nat. Methods</em>, 12(1):59–60, January 2015. <a class="reference external" href="https://doi.org/10.1038/nmeth.3176">doi:10.1038/nmeth.3176</a>.</p>
</dd>
<dt class="label" id="id11"><span class="brackets">CCA+09</span></dt>
<dd><p>Christiam Camacho, George Coulouris, Vahram Avagyan, Ning Ma, Jason Papadopoulos, Kevin Bealer, and Thomas L Madden. BLAST+: architecture and applications. <em>BMC Bioinformatics</em>, 10(1):421, December 2009. <a class="reference external" href="https://doi.org/10.1186/1471-2105-10-421">doi:10.1186/1471-2105-10-421</a>.</p>
</dd>
<dt class="label" id="id15"><span class="brackets">CAK12</span></dt>
<dd><p>Gaudart Corinne, Garrigou Alain, and Chassaing Karine. Analysis of organizational conditions for risk management: the case study of a petrochemical site. <em>Work</em>, 41 Suppl 1:2661–2667, 2012. <a class="reference external" href="https://doi.org/10.3233/wor-2012-1032-2661">doi:10.3233/wor-2012-1032-2661</a>.</p>
</dd>
<dt class="label" id="id38"><span class="brackets">DTCF+17</span></dt>
<dd><p>Paolo Di Tommaso, Maria Chatzou, Evan W Floden, Pablo Prieto Barja, Emilio Palumbo, and Cedric Notredame. Nextflow enables reproducible computational workflows. <em>Nature Biotechnology</em>, 35(4):316–319, 2017. <a class="reference external" href="https://doi.org/10.1038/nbt.3820">doi:10.1038/nbt.3820</a>.</p>
</dd>
<dt class="label" id="id31"><span class="brackets">EMLK16</span></dt>
<dd><p>Philip Ewels, Måns Magnusson, Sverker Lundin, and Max Käller. MultiQC: summarize analysis results for multiple tools and samples in a single report. <em>Bioinformatics</em>, 32(19):3047–3048, 06 2016. URL: <a class="reference external" href="https://doi.org/10.1093/bioinformatics/btw354">https://doi.org/10.1093/bioinformatics/btw354</a>, <a class="reference external" href="https://arxiv.org/abs/https://academic.oup.com/bioinformatics/article-pdf/32/19/3047/25072524/btw354.pdf">arXiv:https://academic.oup.com/bioinformatics/article-pdf/32/19/3047/25072524/btw354.pdf</a>, <a class="reference external" href="https://doi.org/10.1093/bioinformatics/btw354">doi:10.1093/bioinformatics/btw354</a>.</p>
</dd>
<dt class="label" id="id49"><span class="brackets">HEG+20a</span></dt>
<dd><p>Steven Hofmeyr, Rob Egan, Evangelos Georganas, Alex C. Copeland, Robert Riley, Alicia Clum, Emiley Eloe-Fadrosh, Simon Roux, Eugene Goltsman, Aydın Buluç, Daniel Rokhsar, Leonid Oliker, and Katherine Yelick. Terabase-scale metagenome coassembly with metahipmer. <em>Scientific Reports</em>, 10(1):10689, 2020. URL: <a class="reference external" href="https://doi.org/10.1038/s41598-020-67416-5">https://doi.org/10.1038/s41598-020-67416-5</a>, <a class="reference external" href="https://doi.org/10.1038/s41598-020-67416-5">doi:10.1038/s41598-020-67416-5</a>.</p>
</dd>
<dt class="label" id="id48"><span class="brackets">HEG+20b</span></dt>
<dd><p>Steven Hofmeyr, Rob Egan, Evangelos Georganas, Alex C. Copeland, Robert Riley, Alicia Clum, Emiley Eloe-Fadrosh, Simon Roux, Eugene Goltsman, and Aydın et al. Buluç. Terabase-scale metagenome coassembly with metahipmer. <em>Scientific Reports</em>, 2020. <a class="reference external" href="https://doi.org/10.1038/s41598-020-67416-5">doi:10.1038/s41598-020-67416-5</a>.</p>
</dd>
<dt class="label" id="id8"><span class="brackets">HCL+10a</span></dt>
<dd><p>Doug Hyatt, Gwo-Liang Chen, Philip F Locascio, Miriam L Land, Frank W Larimer, and Loren J Hauser. Prodigal: prokaryotic gene recognition and translation initiation site identification. <em>BMC Bioinformatics</em>, 11(1):119, March 2010. <a class="reference external" href="https://doi.org/10.1186/1471-2105-11-119">doi:10.1186/1471-2105-11-119</a>.</p>
</dd>
<dt class="label" id="id47"><span class="brackets">HCL+10b</span></dt>
<dd><p>Doug Hyatt, Gwo-Liang Chen, Philip F LoCascio, Miriam L Land, Frank W Larimer, and Loren J Hauser. Prodigal: prokaryotic gene recognition and translation initiation site identification. <em>BMC Bioinformatics</em>, 2010. <a class="reference external" href="https://doi.org/10.1186/1471-2105-11-119">doi:10.1186/1471-2105-11-119</a>.</p>
</dd>
<dt class="label" id="id23"><span class="brackets">Kan19</span></dt>
<dd><p>M. Kanehisa. Toward understanding the origin and evolution of cellular organisms. <em>Protein Sci</em>, 28(11):1947–1951, 11 2019. <a class="reference external" href="https://doi.org/10.1002/pro.3715">doi:10.1002/pro.3715</a>.</p>
</dd>
<dt class="label" id="id24"><span class="brackets">KFS+21</span></dt>
<dd><p>M. Kanehisa, M. Furumichi, Y. Sato, M. Ishiguro-Watanabe, and M. Tanabe. KEGG: integrating viruses and cellular organisms. <em>Nucleic Acids Res</em>, 49(D1):D545–D551, 01 2021. <a class="reference external" href="https://doi.org/10.1093/nar/gkaa970">doi:10.1093/nar/gkaa970</a>.</p>
</dd>
<dt class="label" id="id22"><span class="brackets">KG00</span></dt>
<dd><p>M. Kanehisa and S. Goto. KEGG: kyoto encyclopedia of genes and genomes. <em>Nucleic Acids Res</em>, 28(1):27–30, Jan 2000. <a class="reference external" href="https://doi.org/10.1093/nar/28.1.27">doi:10.1093/nar/28.1.27</a>.</p>
</dd>
<dt class="label" id="id33"><span class="brackets">KGM16</span></dt>
<dd><p>Kevin P. Keegan, Elizabeth M. Glass, and Folker Meyer. Mg-rast, a metagenomics service for analysis of microbial community structure and function. <em>Microbial Environmental Genomics (MEG)</em>, pages 207–233, 2016. <a class="reference external" href="https://doi.org/10.1007/978-1-4939-3369-3_13">doi:10.1007/978-1-4939-3369-3_13</a>.</p>
</dd>
<dt class="label" id="id35"><span class="brackets">KBZ+20</span></dt>
<dd><p>Silas Kieser, Joseph Brown, Evgeny M. Zdobnov, Mirko Trajkovski, and Lee Ann McCue. Atlas: a snakemake workflow for assembly, annotation, and genomic binning of metagenome sequence data. <em>BMC Bioinformatics</em>, 2020. <a class="reference external" href="https://doi.org/10.1186/s12859-020-03585-4">doi:10.1186/s12859-020-03585-4</a>.</p>
</dd>
<dt class="label" id="id34"><span class="brackets">KSG+22</span></dt>
<dd><p>Sabrina Krakau, Daniel Straub, Hadrien Gourlé, Gisela Gabernet, and Sven Nahnsen. nf-core/mag: a best-practice pipeline for metagenome hybrid assembly and binning. <em>NAR Genomics and Bioinformatics</em>, 02 2022. lqac007. URL: <a class="reference external" href="https://doi.org/10.1093/nargab/lqac007">https://doi.org/10.1093/nargab/lqac007</a>, <a class="reference external" href="https://arxiv.org/abs/https://academic.oup.com/nargab/article-pdf/4/1/lqac007/42366621/lqac007.pdf">arXiv:https://academic.oup.com/nargab/article-pdf/4/1/lqac007/42366621/lqac007.pdf</a>, <a class="reference external" href="https://doi.org/10.1093/nargab/lqac007">doi:10.1093/nargab/lqac007</a>.</p>
</dd>
<dt class="label" id="id26"><span class="brackets">KJE+21</span></dt>
<dd><p>Felix Krueger, Frankie James, Phil Ewels, Ebrahim Afyounian, and Benjamin Schuster-Boeckler. Trimgalore. July 2021. URL: <a class="reference external" href="https://doi.org/10.5281/zenodo.5127899">https://doi.org/10.5281/zenodo.5127899</a>, <a class="reference external" href="https://doi.org/10.5281/zenodo.5127899">doi:10.5281/zenodo.5127899</a>.</p>
</dd>
<dt class="label" id="id39"><span class="brackets">KSB17</span></dt>
<dd><p>Gregory M. Kurtzer, Vanessa Sochat, and Michael W. Bauer. Singularity: scientific containers for mobility of compute. <em>PLOS ONE</em>, 12(5):e0177459, 2017. <a class="reference external" href="https://doi.org/10.1371/journal.pone.0177459">doi:10.1371/journal.pone.0177459</a>.</p>
</dd>
<dt class="label" id="id4"><span class="brackets">LS12</span></dt>
<dd><p>Ben Langmead and Steven L Salzberg. Fast gapped-read alignment with Bowtie 2. <em>Nature Methods</em>, 9(4):357–359, 2012. URL: <a class="reference external" href="https://doi.org/10.1038/nmeth.1923">https://doi.org/10.1038/nmeth.1923</a>, <a class="reference external" href="https://doi.org/10.1038/nmeth.1923">doi:10.1038/nmeth.1923</a>.</p>
</dd>
<dt class="label" id="id46"><span class="brackets">LLL+15a</span></dt>
<dd><p>Dinghua Li, Chi-Man Liu, Ruibang Luo, Kunihiko Sadakane, and Tak-Wah Lam. Megahit: an ultra-fast single-node solution for large and complex metagenomics assembly via succinct de bruijn graph. <em>Bioinformatics</em>, 31(10):1674–1676, 2015. <a class="reference external" href="https://doi.org/10.1093/bioinformatics/btv033">doi:10.1093/bioinformatics/btv033</a>.</p>
</dd>
<dt class="label" id="id20"><span class="brackets">LLL+15b</span></dt>
<dd><p>Dinghua Li, Chi-Man Liu, Ruibang Luo, Kunihiko Sadakane, and Tak-Wah Lam. MEGAHIT: an ultra-fast single-node solution for large and complex metagenomics assembly via succinct de Bruijn graph. <em>Bioinformatics</em>, 31(10):1674–1676, 01 2015. URL: <a class="reference external" href="https://doi.org/10.1093/bioinformatics/btv033">https://doi.org/10.1093/bioinformatics/btv033</a>, <a class="reference external" href="https://arxiv.org/abs/https://academic.oup.com/bioinformatics/article-pdf/31/10/1674/17085710/btv033.pdf">arXiv:https://academic.oup.com/bioinformatics/article-pdf/31/10/1674/17085710/btv033.pdf</a>, <a class="reference external" href="https://doi.org/10.1093/bioinformatics/btv033">doi:10.1093/bioinformatics/btv033</a>.</p>
</dd>
<dt class="label" id="id10"><span class="brackets">LD10</span></dt>
<dd><p>Heng Li and Richard Durbin. Fast and accurate long-read alignment with Burrows-Wheeler transform. <em>Bioinformatics</em>, 26(5):589–595, March 2010. <a class="reference external" href="https://doi.org/10.1093/bioinformatics/btp698">doi:10.1093/bioinformatics/btp698</a>.</p>
</dd>
<dt class="label" id="id30"><span class="brackets">LHW+09</span></dt>
<dd><p>Heng Li, Bob Handsaker, Alec Wysoker, Tim Fennell, Jue Ruan, Nils Homer, Gabor Marth, Goncalo Abecasis, Richard Durbin, and 1000 Genome Project Data Processing Subgroup. The Sequence Alignment/Map format and SAMtools. <em>Bioinformatics</em>, 25(16):2078–2079, 06 2009. URL: <a class="reference external" href="https://doi.org/10.1093/bioinformatics/btp352">https://doi.org/10.1093/bioinformatics/btp352</a>, <a class="reference external" href="https://arxiv.org/abs/https://academic.oup.com/bioinformatics/article-pdf/25/16/2078/531810/btp352.pdf">arXiv:https://academic.oup.com/bioinformatics/article-pdf/25/16/2078/531810/btp352.pdf</a>, <a class="reference external" href="https://doi.org/10.1093/bioinformatics/btp352">doi:10.1093/bioinformatics/btp352</a>.</p>
</dd>
<dt class="label" id="id6"><span class="brackets">MagovcS11</span></dt>
<dd><p>Tanja Magoč and Steven L Salzberg. FLASH: fast length adjustment of short reads to improve genome assemblies. <em>Bioinformatics</em>, 27(21):2957–2963, November 2011. <a class="reference external" href="https://doi.org/10.1093/bioinformatics/btr507">doi:10.1093/bioinformatics/btr507</a>.</p>
</dd>
<dt class="label" id="id40"><span class="brackets">MSB+21a</span></dt>
<dd><p>M Mirdita, M Steinegger, F Breitwieser, J Söding, and E Levy Karin. Fast and sensitive taxonomic assignment to metagenomic contigs. <em>Bioinformatics</em>, 37(18):3029–3031, 2021. <a class="reference external" href="https://doi.org/10.1093/bioinformatics/btab184">doi:10.1093/bioinformatics/btab184</a>.</p>
</dd>
<dt class="label" id="id32"><span class="brackets">MSB+21b</span></dt>
<dd><p>M Mirdita, M Steinegger, F Breitwieser, J Söding, and E Levy Karin. Fast and sensitive taxonomic assignment to metagenomic contigs. <em>Bioinformatics</em>, 37(18):3029–3031, 03 2021. URL: <a class="reference external" href="https://doi.org/10.1093/bioinformatics/btab184">https://doi.org/10.1093/bioinformatics/btab184</a>, <a class="reference external" href="https://arxiv.org/abs/https://academic.oup.com/bioinformatics/article-pdf/37/18/3029/40471478/btab184.pdf">arXiv:https://academic.oup.com/bioinformatics/article-pdf/37/18/3029/40471478/btab184.pdf</a>, <a class="reference external" href="https://doi.org/10.1093/bioinformatics/btab184">doi:10.1093/bioinformatics/btab184</a>.</p>
</dd>
<dt class="label" id="id37"><span class="brackets">MJL+21</span></dt>
<dd><p>Felix Mölder, Kim Philipp Jablonski, Brice Letcher, Michael B. Hall, Christopher H. Tomkins-Tinch, Vanessa Sochat, Jan Forster, Soohyun Lee, Sven O. Twardziok, and Alexander et al. Kanitz. Sustainable data analysis with snakemake. <em>F1000Research</em>, 10:33, 2021. <a class="reference external" href="https://doi.org/10.12688/f1000research.29032.1">doi:10.12688/f1000research.29032.1</a>.</p>
</dd>
<dt class="label" id="id45"><span class="brackets">PLYC12a</span></dt>
<dd><p>Y. Peng, H. C. M. Leung, S. M. Yiu, and F. Y. L. Chin. Idba-ud: a de novo assembler for single-cell and metagenomic sequencing data with highly uneven depth. <em>Bioinformatics</em>, 28(11):1420–1428, 2012. <a class="reference external" href="https://doi.org/10.1093/bioinformatics/bts174">doi:10.1093/bioinformatics/bts174</a>.</p>
</dd>
<dt class="label" id="id21"><span class="brackets">PLYC12b</span></dt>
<dd><p>Yu Peng, Henry C. M. Leung, S. M. Yiu, and Francis Y. L. Chin. IDBA-UD: a de novo assembler for single-cell and metagenomic sequencing data with highly uneven depth. <em>Bioinformatics</em>, 28(11):1420–1428, 04 2012. URL: <a class="reference external" href="https://doi.org/10.1093/bioinformatics/bts174">https://doi.org/10.1093/bioinformatics/bts174</a>, <a class="reference external" href="https://arxiv.org/abs/https://academic.oup.com/bioinformatics/article-pdf/28/11/1420/742285/bts174.pdf">arXiv:https://academic.oup.com/bioinformatics/article-pdf/28/11/1420/742285/bts174.pdf</a>, <a class="reference external" href="https://doi.org/10.1093/bioinformatics/bts174">doi:10.1093/bioinformatics/bts174</a>.</p>
</dd>
<dt class="label" id="id3"><span class="brackets">RLT+18</span></dt>
<dd><p>Ben Roediger, Quintin Lee, Shweta Tikoo, Joanna C A Cobbin, James M Henderson, Mika Jormakka, Matthew B O'Rourke, Matthew P Padula, Natalia Pinello, Marisa Henry, Maria Wynne, Sara F Santagostino, Cory F Brayton, Lorna Rasmussen, Leszek Lisowski, Szun S Tay, David C Harris, John F Bertram, John P Dowling, Patrick Bertolino, Jack H Lai, Wengen Wu, William W Bachovchin, Justin J-L Wong, Mark D Gorrell, Babak Shaban, Edward C Holmes, Christopher J Jolly, Sébastien Monette, and Wolfgang Weninger. An atypical Parvovirus drives chronic tubulointerstitial nephropathy and kidney fibrosis. <em>Cell</em>, 175(2):530–543.e24, October 2018. <a class="reference external" href="https://doi.org/10.1016/j.cell.2018.08.013">doi:10.1016/j.cell.2018.08.013</a>.</p>
</dd>
<dt class="label" id="id19"><span class="brackets">SGS19</span></dt>
<dd><p>Erika Sallet, Jérôme Gouzy, and Thomas Schiex. <em>EuGene: An Automated Integrative Gene Finder for Eukaryotes and Prokaryotes</em>, pages 97–120. Springer New York, New York, NY, 2019. URL: <a class="reference external" href="https://doi.org/10.1007/978-1-4939-9173-0_6">https://doi.org/10.1007/978-1-4939-9173-0_6</a>, <a class="reference external" href="https://doi.org/10.1007/978-1-4939-9173-0_6">doi:10.1007/978-1-4939-9173-0_6</a>.</p>
</dd>
<dt class="label" id="id7"><span class="brackets">SE11</span></dt>
<dd><p>Robert Schmieder and Robert Edwards. Fast identification and removal of sequence contamination from genomic and metagenomic datasets. <em>PLoS One</em>, 6(3):e17288, March 2011. <a class="reference external" href="https://doi.org/10.1371/journal.pone.0017288">doi:10.1371/journal.pone.0017288</a>.</p>
</dd>
<dt class="label" id="id13"><span class="brackets">SeS12</span></dt>
<dd><p>Marcello Silva e Santos. The PhOCoe model–ergonomic pattern mapping in participatory design processes. <em>Work</em>, 41 Suppl 1:2643–2650, 2012. <a class="reference external" href="https://doi.org/10.3233/WOR-2012-0507-2643">doi:10.3233/WOR-2012-0507-2643</a>.</p>
</dd>
<dt class="label" id="id14"><span class="brackets">TLevequeD+12</span></dt>
<dd><p>Michèle Tosello, Françoise Lévêque, Stéphanie Dutillieu, Guillaume Hernandez, and Jean-François Vautier. Conditions for the successful integration of human and organizational factors (HOF) in the nuclear safety analysis. <em>Work</em>, 41 Suppl 1:2656–2660, 2012. <a class="reference external" href="https://doi.org/10.3233/wor-2012-0508-2656">doi:10.3233/wor-2012-0508-2656</a>.</p>
</dd>
<dt class="label" id="id43"><span class="brackets">UC18</span></dt>
<dd><p>The UniProt Consortium. Uniprot: the universal protein knowledgebase. <em>Nucleic Acids Research</em>, 46(5):2699–2699, 2018. <a class="reference external" href="https://doi.org/10.1093/nar/gky092">doi:10.1093/nar/gky092</a>.</p>
</dd>
<dt class="label" id="id36"><span class="brackets">VDHV+21</span></dt>
<dd><p>Renaud Van Damme, Martin Hölzer, Adrian Viehweger, Bettina Müller, Erik Bongcam-Rudloff, and Christian Brandt. Metagenomics workflow for hybrid assembly, differential coverage binning, metatranscriptomics and pathway analysis (muffin). <em>PLOS Computational Biology</em>, 17(2):e1008716, 2021. <a class="reference external" href="https://doi.org/10.1371/journal.pcbi.1008716">doi:10.1371/journal.pcbi.1008716</a>.</p>
</dd>
<dt class="label" id="id41"><span class="brackets">VVdAG22</span></dt>
<dd><p>Kate Voss, Geraldine Van der Auwera, and Jeff Gentry. Full-stack genomics pipelining with gatk4 + wdl + cromwell. 2022. URL: <a class="reference external" href="https://f1000research.com/slides/6-1381">https://f1000research.com/slides/6-1381</a>, <a class="reference external" href="https://doi.org/10.7490/f1000research.1114634.1">doi:10.7490/f1000research.1114634.1</a>.</p>
</dd>
</dl>
</div>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="citation.html" class="btn btn-neutral float-left" title="Citation and Attribution" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="contributing.html" class="btn btn-neutral float-right" title="Contributing" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, Babak Shaban, Maria del Mar Quiroga, Edoardo Tescari, Robert Turnbull, Heroen Verbruggen, Kim-Anh Lê Cao.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>