---
title: 'MetaGenePipe: A pipeline for metagenomics'
tags:
  - metagenomics
authors:
  - name: Babak Shaban
    orcid: 0000-0002-7393-810X
    affiliation: 1
  - name: Maria del Mar Quiroga
    orcid: 0000-0002-8943-2808
    affiliation: 1
  - name: Edoardo Tescari
    orcid: 0000-0003-1157-4897
    affiliation: 1
  - name: Robert Turnbull
    orcid: 0000-0003-1274-6750
    affiliation: 1
  - name: Heroen Verbruggen
    orcid: 0000-0002-6305-4749
    affiliation: 2
  - name: Kim-Anh LÃª Cao
    orcid: 0000-0003-3923-1116
    affiliation: 3
affiliations:
 - name: Melbourne Data Analytics Platform, University of Melbourne
   index: 1
 - name: School of BioSciences, University of Melbourne
   index: 2
 - name: School of Mathematics and Statistics, Melbourne Integrative Genomics, University of Melbourne
   index: 3
date: 16 March 2022
bibliography: docs/refs.bib

---

# Summary

Microorganisms including bacteria, viruses, archaea, fungi are ubiquitous in our environment. The study of microorganisms and their full genomes has been enabled through advances in culture independent techniques and high-throughput sequencing technologies. Whole genome metagenomics shotgun sequencing (WGS) empowers researchers to study biological functions of microorganisms, and how their presence affect human disease or a specific ecosystem. However, advanced and novel bioinformatics techniques are required to process the data into a suitable format. There is no standardised bioinformatics framework a microbiologist can use effectively.

# Statement of need



# Acknowledgements



# References

[@Arden2017-as]
[@Roediger2018-lq]
[@Andrews:2010tn]
[@Magoc2011-gb]
[@Schmieder2011-jr]
[@Hyatt2010-zh]
[@Buchfink2015-rn]
[@Li2010-nl]
[@Camacho2009-hf]
[@Altschul1997-oe]
[@Silva_e_Santos2012-qm]
[@Tosello2012-kp]
[@Corinne2012-tx]
[@Herrera2012-qm]
[@Soares2012-us]
[@Blome2012-dx]
[@Bitencourt2012-vu]
[@Coelho2012-tk]
[@Scariot2012-bl]
[@Zulch2012-ul]
[@Altschul1990-xn]

