## MetaGenePipe developed at Melbourne Data Analytics Platform.

Please refer to documentation for how to run.

```
"metaGenPipe.DB": "kegg/kegg.dmnd",
```

## Citation and Attribution

We currently are authoring a paper for the Journal of Open Source Software about this software package. Citation details will be added upon publication.

If you create a derivative work from this software package, attribution should be included as follows:

> This is a derivative work of MetaGenePipe, originally released under the Apache 2.0 license, developed by Bobbie Shaban, Mar Quiroga, Robert Turnbull and Edoardo Tescari at Melbourne Data Analytics Platform (MDAP) at the University of Melbourne.

## Troubleshooting tips:
` The pipeline has been set up to run against the swissprot database. We have supplied sample fastq files consisting of 100,000 reads so the pipeline can be tested.`

` If you would like access to the kegg database, send an email to bshaban@unimelb.edu.au and he may be able to see if you're eligble to use it. `


**NOTE: Some reference files have not been added due to be kegg reference files. These must be obtained separately with authorisation**

## References for software used

* [1]: Detection of Toscana virus from an adult traveler returning to Australia with encephalitis. Katherine E. Arden  Claire Heney  Babak Shaban  Graeme R. Nimmo  Michael D. Nissen  Theo P. Sloots  Ian M. Mackay.https://doi.org/10.1002/jmv.24839

* [2]: An atypical parvovirus drives chronic tubulointerstitial nephropathy and kidney fibrosis. B Roediger, Q Lee, S Tikoo, JCA Cobbin, JM Henderson, M Jormakka, Babak Shaban. Cell 175 (2), 530-543. e24

* Andrews S. (2010). FastQC: a quality control tool for high throughput sequence data. Available online at: http://www.bioinformatics.babraham.ac.uk/projects/fastqc

* FLASH: Fast length adjustment of short reads to improve genome assemblies. T. Magoc and S. Salzberg. Bioinformatics 27:21 (2011), 2957-63.

* Deconseq: Schmieder R and Edwards R: Fast identification and removal of sequence contamination from genomic and metagenomic datasets. PLoS ONE 2011, 6:e17288. [PMID: 21408061]

* Prodigal: prokaryotic gene recognition and translation initiation site identification: Doug Hyatt, Gwo-Liang Chen,1 Philip F LoCascio,1 Miriam L Land,1,3 Frank W Larimer,1,2 and Loren J Hauser1,3

* Buchfink B, Xie C, Huson DH, "Fast and sensitive protein alignment using DIAMOND", Nature Methods 12, 59-60 (2015). doi:10.1038/nmeth.3176

* Li H. and Durbin R. (2010) Fast and accurate long-read alignment with Burrows-Wheeler transform. Bioinformatics, 26, 589-595. [PMID: 20080505]

*  Camacho C. et al. (2009) "BLAST+: architecture and applications"
*  Altschul S.F. et al. (1997) "Gapped BLAST and PSI-BLAST: a new generation of protein database search programs"
*  Altschul S.F. et al. (1990) "Basic local alignment search tool"



