<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Workflow &mdash; MetaGenePipe 1.0.0 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="_static/copybutton.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/clipboard.min.js"></script>
        <script src="_static/copybutton.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Extension" href="extension.html" />
    <link rel="prev" title="Usage" href="usage.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="index.html" class="icon icon-home"> MetaGenePipe
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="usage.html">Usage</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Workflow</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#output">Output</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#quality-control">Quality Control</a></li>
<li class="toctree-l3"><a class="reference internal" href="#data">Data</a></li>
<li class="toctree-l3"><a class="reference internal" href="#assembly">Assembly</a></li>
<li class="toctree-l3"><a class="reference internal" href="#read-alignment">Read alignment</a></li>
<li class="toctree-l3"><a class="reference internal" href="#gene-prediction">Gene prediction</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#output-tree">Output Tree</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="extension.html">Extension</a></li>
<li class="toctree-l1"><a class="reference internal" href="citation.html">Citation and Attribution</a></li>
<li class="toctree-l1"><a class="reference internal" href="bibliography.html">Bibliography</a></li>
<li class="toctree-l1"><a class="reference internal" href="contributing.html">Contributing</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">MetaGenePipe</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
      <li>Workflow</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/workflow.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="workflow">
<h1>Workflow<a class="headerlink" href="#workflow" title="Permalink to this headline"></a></h1>
<img alt="MetaGenePipe Workflow" src="_images/MetaGenePipe.drawio.svg" /><ul class="simple">
<li><p>Flash: Lengthens reads</p></li>
<li><p>Blast: Aligns contigs to Blast database</p></li>
<li><p>Merge: Concatenates samples</p></li>
<li><p>Taxon: Performs taxonomic assignment</p></li>
<li><p>Trim Galore: Performs quality control for reads</p></li>
<li><p>Read Alignment: Aligns reads to contigs</p></li>
<li><p>HMMER: Performs brite analysis</p></li>
</ul>
<p>The submodules of the workflow are configured in <code class="docutils literal notranslate"><span class="pre">metaGenePipe.json</span></code>. The defaults are as follows:</p>
<div class="highlight-json notranslate"><div class="highlight"><pre><span></span><span class="nt">&quot;metaGenPipe.flashBoolean&quot;</span><span class="p">:</span><span class="w"> </span><span class="kc">true</span><span class="p">,</span><span class="w"></span>
<span class="nt">&quot;metaGenPipe.blastBoolean&quot;</span><span class="p">:</span><span class="w"> </span><span class="kc">false</span><span class="p">,</span><span class="w"></span>
<span class="nt">&quot;metaGenPipe.concatenateBoolean&quot;</span><span class="p">:</span><span class="w"> </span><span class="kc">true</span><span class="p">,</span><span class="w"></span>
<span class="nt">&quot;metaGenPipe.taxonBoolean&quot;</span><span class="p">:</span><span class="w"> </span><span class="kc">true</span><span class="p">,</span><span class="w"></span>
<span class="nt">&quot;metaGenPipe.trimmomaticBoolean&quot;</span><span class="p">:</span><span class="w"> </span><span class="kc">false</span><span class="p">,</span><span class="w"></span>
<span class="nt">&quot;metaGenPipe.trimGaloreBoolean&quot;</span><span class="p">:</span><span class="w"> </span><span class="kc">true</span><span class="p">,</span><span class="w"></span>
<span class="nt">&quot;metaGenPipe.megahitBoolean&quot;</span><span class="p">:</span><span class="w"> </span><span class="kc">true</span><span class="p">,</span><span class="w"></span>
<span class="nt">&quot;metaGenPipe.mapreadsBoolean&quot;</span><span class="p">:</span><span class="w"> </span><span class="kc">true</span><span class="p">,</span><span class="w"></span>
<span class="nt">&quot;metaGenPipe.hmmerBoolean&quot;</span><span class="p">:</span><span class="w"> </span><span class="kc">true</span><span class="p">,</span><span class="w"></span>
</pre></div>
</div>
<section id="output">
<h2>Output<a class="headerlink" href="#output" title="Permalink to this headline"></a></h2>
<p>There are four main output folders: qc (quality control), assembly, readalignment, and geneprediction and one intermediary, data, which contains the samples for assembly after running through TrimGalore and concatenating the samples for co-assembly if specified.</p>
<p>An example tree of the the output directory and the associated output definitions are below:</p>
<section id="quality-control">
<h3>Quality Control<a class="headerlink" href="#quality-control" title="Permalink to this headline"></a></h3>
<ul class="simple">
<li><p>trimmed</p>
<ul>
<li><p>{sampleName}.T{G|T}_R{1|2}.fq.gz: Trimmed output for each of the individual sample files, TG if the chosen trimmer is TrimGalore, and TT if it is Trimmomatic</p></li>
</ul>
</li>
<li><p>fastqc</p>
<ul>
<li><p>{sampleName}.T{G|T}_R{1|2}_fastqc.zip: Fastqc output for each of the individual sample files</p></li>
</ul>
</li>
<li><p>multiqc_report.html: Combined report of all fastqc files</p></li>
<li><p>flash</p>
<ul>
<li><p>{sampleName}.extendedFrags.fastq: [ADD HERE]</p></li>
</ul>
</li>
</ul>
</section>
<section id="data">
<h3>Data<a class="headerlink" href="#data" title="Permalink to this headline"></a></h3>
<ul class="simple">
<li><p>{sampleName}_R{1|2}.fq.gz Sample files after trimming and/or concatenating for co-assembly. If files are concatenated for co-assembly, the sample name is set to be <cite>combined</cite></p></li>
</ul>
</section>
<section id="assembly">
<h3>Assembly<a class="headerlink" href="#assembly" title="Permalink to this headline"></a></h3>
<ul class="simple">
<li><p>{sampleName}.megahit.contigs.fa: Final assembled contigs</p></li>
<li><p>{sampleName}.{kmer}.fastg: Assembly graph for {kmer} assembled contigs, where {kmer} produces the largest assembled contig file size in the <cite>intermediate_contigs</cite> folder</p></li>
<li><p>intermediate_contigs: a folder containing all intermediate assembled contigs {sampleName}.contigs.k{kmer}.fastg</p></li>
<li><p>{sampleName}.megahit.blast.out: Raw blast results for the contigs</p></li>
<li><p>{sampleName}.megahit.blast.parsed: Blast results parsed to be easily viewed in tsv format</p></li>
</ul>
</section>
<section id="read-alignment">
<h3>Read alignment<a class="headerlink" href="#read-alignment" title="Permalink to this headline"></a></h3>
<ul class="simple">
<li><p>{sampleName}.T{G|T}.flagstat.txt: Samtools flagstat output. Reports statistics on alignment of reads back to assembled contigs</p></li>
<li><p>{sampleName}.T{G|T}.sam: Alignment of reads back to contigs in SAM format</p></li>
<li><p>{sampleName}.T{G|T}.sorted.bam: Alignment of reads back to contigs in BAM format</p></li>
</ul>
</section>
<section id="gene-prediction">
<h3>Gene prediction<a class="headerlink" href="#gene-prediction" title="Permalink to this headline"></a></h3>
<ul class="simple">
<li><p>{sampleName}.megahit.proteins.fa.xml.out.xml: XML output of alignment of predicted Amino Acids to NCBI database (We chose swissprot, but any blast database can be substituted)</p></li>
<li><p>diamond</p>
<ul>
<li><p>{sampleName}.megahit.proteins.fa.xml.out:</p></li>
</ul>
</li>
<li><p>hmmer</p>
<ul>
<li><p>combined.megahit.proteins.hmmer.out: Raw hmmer output aligned to Koalafam profiles</p></li>
<li><p>combined.megahit.proteins.hmmer.tblout: Parsed hmmer output aligned to Koalafam profiles</p></li>
</ul>
</li>
<li><p>prodigal</p>
<ul>
<li><p>combined.megahit.gene_coordinates.gbk: Gene coordinates file (Genbank like file)</p></li>
<li><p>combined.megahit.nucl_genes.fa: Predicted gene nucleotide sequences</p></li>
<li><p>combined.megahit.proteins.fa: Predicted gene amino acid sequences</p></li>
<li><p>combined.megahit.starts.txt: Prodigal starts file</p></li>
</ul>
</li>
<li><p>taxon</p>
<ul>
<li><p>LevelA.brite.counts.tsv: Level A Kegg Brite Hierarchical count</p></li>
<li><p>LevelB.brite.counts.tsv: Level B Kegg Brite Hierarchical count</p></li>
<li><p>LevelC.brite.counts.tsv: Level C Kegg Brite Hierarchical count</p></li>
<li><p>OTU.brite.tsv: [DESCRIBE HERE]</p></li>
</ul>
</li>
</ul>
<section id="output-tree">
<h4>Output Tree<a class="headerlink" href="#output-tree" title="Permalink to this headline"></a></h4>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>.
├── assembly
│   ├── combined.57.fastg
│   ├── combined.megahit.blast.out
│   ├── combined.megahit.blast.parsed
│   ├── combined.megahit.contigs.fa
│   └── intermediate_contigs
│       ├── combined.contigs.k27.fa
│       ├── combined.contigs.k37.fa
│       ├── combined.contigs.k47.fa
│       ├── combined.contigs.k57.fa
│       ├── combined.contigs.k67.fa
│       ├── combined.contigs.k77.fa
│       ├── combined.contigs.k87.fa
│       └── combined.contigs.k97.fa
├── data
│   ├── combined_R1.fq.gz
│   └── combined_R2.fq.gz
├── geneprediction
│   ├── combined.megahit.proteins.fa.xml.out.xml
│   ├── diamond
│   │   └── combined.megahit.proteins.fa.xml.out
│   ├── hmmer
│   │   ├── combined.megahit.proteins.hmmer.out
│   │   └── combined.megahit.proteins.hmmer.tblout
│   ├── prodigal
│   │   ├── combined.megahit.gene_coordinates.gbk
│   │   ├── combined.megahit.nucl_genes.fa
│   │   ├── combined.megahit.proteins.fa
│   │   └── combined.megahit.starts.txt
│   └── taxon
│       ├── Level1.brite.counts.tsv
│       ├── Level2.brite.counts.tsv
│       ├── Level3.brite.counts.tsv
│       └── OTU.brite.tsv
├── qc
│   ├── fastqc
│   │   ├── SRR5808831.TG_R1_fastqc.zip
│   │   ├── SRR5808831.TG_R2_fastqc.zip
│   │   ├── SRR5808882.TG_R1_fastqc.zip
│   │   └── SRR5808882.TG_R2_fastqc.zip
│   ├── flash
│   │   ├── SRR5808831.extendedFrags.fastq
│   │   └── SRR5808882.extendedFrags.fastq
│   ├── multiqc_report.html
│   └── trimmed
│       ├── SRR5808831.TG_R1.fq.gz
│       ├── SRR5808831.TG_R2.fq.gz
│       ├── SRR5808882.TG_R1.fq.gz
│       └── SRR5808882.TG_R2.fq.gz
└── readalignment
   ├── SRR5808831.TG.flagstat.txt
   ├── SRR5808831.TG.sam
   ├── SRR5808831.TG.sorted.bam
   ├── SRR5808882.TG.flagstat.txt
   ├── SRR5808882.TG.sam
   └── SRR5808882.TG.sorted.bam
</pre></div>
</div>
</section>
</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="usage.html" class="btn btn-neutral float-left" title="Usage" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="extension.html" class="btn btn-neutral float-right" title="Extension" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, Babak Shaban, Maria del Mar Quiroga, Edoardo Tescari, Robert Turnbull, Heroen Verbruggen, Kim-Anh Lê Cao.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>