<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Workflow &mdash; MetaGenePipe 1.0.0 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="_static/copybutton.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/clipboard.min.js"></script>
        <script src="_static/copybutton.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Extension" href="extension.html" />
    <link rel="prev" title="Usage" href="usage.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="index.html" class="icon icon-home"> MetaGenePipe
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="usage.html">Usage</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Workflow</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#output">Output</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#output-tree">Output Tree</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="extension.html">Extension</a></li>
<li class="toctree-l1"><a class="reference internal" href="citation.html">Citation and Attribution</a></li>
<li class="toctree-l1"><a class="reference internal" href="bibliography.html">Bibliography</a></li>
<li class="toctree-l1"><a class="reference internal" href="contributing.html">Contributing</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">MetaGenePipe</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
      <li>Workflow</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/workflow.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="workflow">
<h1>Workflow<a class="headerlink" href="#workflow" title="Permalink to this headline"></a></h1>
<img alt="MetaGenePipe Workflow" src="_images/MetaGenePipe.drawio.svg" /><ul class="simple">
<li><p>Flash: Lengthens reads</p></li>
<li><p>Blast: Aligns contigs to Blast database</p></li>
<li><p>Merge: Concatenates samples</p></li>
<li><p>Taxon: Performs taxonomic assignment</p></li>
<li><p>Trim Galore: Performs quality control for reads</p></li>
<li><p>Read Alignment: Aligns reads to contigs</p></li>
<li><p>HMMER: Performs brite analysis</p></li>
</ul>
<p>The submodules of the workflow are configured in <code class="docutils literal notranslate"><span class="pre">metaGenePipe.json</span></code>. The defaults are as follows:</p>
<div class="highlight-json notranslate"><div class="highlight"><pre><span></span><span class="nt">&quot;metaGenPipe.flashBoolean&quot;</span><span class="p">:</span><span class="w"> </span><span class="kc">true</span><span class="p">,</span><span class="w"></span>
<span class="nt">&quot;metaGenPipe.blastBoolean&quot;</span><span class="p">:</span><span class="w"> </span><span class="kc">false</span><span class="p">,</span><span class="w"></span>
<span class="nt">&quot;metaGenPipe.concatenateBoolean&quot;</span><span class="p">:</span><span class="w"> </span><span class="kc">true</span><span class="p">,</span><span class="w"></span>
<span class="nt">&quot;metaGenPipe.taxonBoolean&quot;</span><span class="p">:</span><span class="w"> </span><span class="kc">true</span><span class="p">,</span><span class="w"></span>
<span class="nt">&quot;metaGenPipe.trimmomaticBoolean&quot;</span><span class="p">:</span><span class="w"> </span><span class="kc">false</span><span class="p">,</span><span class="w"></span>
<span class="nt">&quot;metaGenPipe.trimGaloreBoolean&quot;</span><span class="p">:</span><span class="w"> </span><span class="kc">true</span><span class="p">,</span><span class="w"></span>
<span class="nt">&quot;metaGenPipe.megahitBoolean&quot;</span><span class="p">:</span><span class="w"> </span><span class="kc">true</span><span class="p">,</span><span class="w"></span>
<span class="nt">&quot;metaGenPipe.mapreadsBoolean&quot;</span><span class="p">:</span><span class="w"> </span><span class="kc">true</span><span class="p">,</span><span class="w"></span>
<span class="nt">&quot;metaGenPipe.hmmerBoolean&quot;</span><span class="p">:</span><span class="w"> </span><span class="kc">true</span><span class="p">,</span><span class="w"></span>
</pre></div>
</div>
<section id="output">
<h2>Output<a class="headerlink" href="#output" title="Permalink to this headline"></a></h2>
<p>There are five main folders of output (Assembly, Gene Prediction, Read Alignment, QC and Taxon) and one intermediary (data) which contains the merging of raw samples and the output from Trim Galore.</p>
<p>An example tree of the the output directory and the associated output definitions are below:</p>
<p>The Assembly directory contains the following</p>
<ul class="simple">
<li><p>merged.contigs.k27.fa: Kmer assembled contigs: assembled contigs for the kmer values, represented in the “intermediate_contigs” folder</p></li>
<li><p>merged.megahit.contigs.fa: Final assembled contigs</p></li>
<li><p>merged.37.fastg: A fastg file. Fastg is the assembly graph produced by the assembler.</p></li>
<li><p>merged.megahit.blast.out: Raw blast results for the contigs</p></li>
<li><p>merged.megahit.blast.parsed: Blast results parsed to be easily viewed in tsv format</p></li>
</ul>
<p>Gene prediction contains the output from prodigal</p>
<ul class="simple">
<li><p>Merge.hmmer.out: Raw hmmer output aligned to Koalafam profiles</p></li>
<li><p>Merge.hmmer.tblout: Parsed hmmer output aligned to Koalafam profiles</p></li>
<li><p>Merge.prodigal.genes.fa: Gene coordinates file (Genbank like file)</p></li>
<li><p>Merge.prodigal.nucl.genes.fa: Predicted gene nucleotide sequences</p></li>
<li><p>Merge.prodigal.potential_genes.fa: Prodigal starts file</p></li>
<li><p>Merge.prodigal.proteins.fa: Predicted gene amino acid sequences</p></li>
<li><p>Merge.xml: XML output of alignment of predicted Amino Acids to NCBI database (We chose swissprot, but any blast database can be substituted)</p></li>
</ul>
<p>Quality Control</p>
<ul class="simple">
<li><p>SRR5808831.TG_R1_fastqc.zip: Fastqc output for each of the individual sample files</p></li>
<li><p>multiqc_report.html: Combined report of all fastqc files</p></li>
</ul>
<p>Read Alignment</p>
<ul class="simple">
<li><p>SRR5808831.TG.flagstat.txt: Samtools flagstat output. Reports statistics on alignment of reads back to assembled contigs</p></li>
<li><p>SRR5808831.TG.sam: Alignment of reads back to contigs in SAM format</p></li>
<li><p>SRR5808831.TG.sorted.bam: Alignment of reads back to contigs in BAM format</p></li>
</ul>
<p>Taxon output</p>
<ul class="simple">
<li><p>Level1.brite.counts.tsv: Level 1 Kegg Brite Heirarchical count</p></li>
<li><p>Level2.brite.counts.tsv: Level 2 Kegg Brite Heirarchical count</p></li>
<li><p>Level3.brite.counts.tsv: Level 3 Kegg Brite Heirarchical count</p></li>
</ul>
<section id="output-tree">
<h3>Output Tree<a class="headerlink" href="#output-tree" title="Permalink to this headline"></a></h3>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>.
├── assembly
│   ├── intermediate_contigs
│   │   ├── merged.contigs.k27.fa
│   │   ├── merged.contigs.k37.fa
│   │   ├── merged.contigs.k47.fa
│   │   ├── merged.contigs.k57.fa
│   │   ├── merged.contigs.k67.fa
│   │   ├── merged.contigs.k77.fa
│   │   ├── merged.contigs.k87.fa
│   │   └── merged.contigs.k97.fa
│   ├── merged.37.fastg
│   ├── merged.megahit.blast.out
│   ├── merged.megahit.blast.parsed
│   └── merged.megahit.contigs.fa
├── data
│   ├── merged
│   │   ├── merged_R1.fq.gz
│   │   └── merged_R2.fq.gz
│   └── trimmed
│       ├── SRR5808831.TG_R1.fq.gz
│       ├── SRR5808831.TG_R2.fq.gz
│       ├── SRR5808882.TG_R1.fq.gz
│       └── SRR5808882.TG_R2.fq.gz
├── geneprediction
│   ├── Merge.hmmer.out
│   ├── Merge.hmmer.tblout
│   ├── Merge.prodigal.genes.fa
│   ├── Merge.prodigal.nucl.genes.fa
│   ├── Merge.prodigal.potential_genes.fa
│   ├── Merge.prodigal.proteins.fa
│   ├── Merge.xml
│   └── Merge.xml.out
├── qc
│   ├── fastqc_zip
│   │   ├── SRR5808831.TG_R1_fastqc.zip
│   │   ├── SRR5808831.TG_R2_fastqc.zip
│   │   ├── SRR5808882.TG_R1_fastqc.zip
│   │   └── SRR5808882.TG_R2_fastqc.zip
│   └── multiqc_report.html
├── readalignment
│   ├── SRR5808831.TG.flagstat.txt
│   ├── SRR5808831.TG.sam
│   ├── SRR5808831.TG.sorted.bam
│   ├── SRR5808882.TG.flagstat.txt
│   ├── SRR5808882.TG.sam
│   └── SRR5808882.TG.sorted.bam
├── SRR5808831.extendedFrags.fastq
├── SRR5808882.extendedFrags.fastq
└── taxon
    ├── Level1.brite.counts.tsv
    ├── Level2.brite.counts.tsv
    └── Level3.brite.counts.tsv
</pre></div>
</div>
</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="usage.html" class="btn btn-neutral float-left" title="Usage" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="extension.html" class="btn btn-neutral float-right" title="Extension" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, Babak Shaban, Maria del Mar Quiroga, Edoardo Tescari, Robert Turnbull, Heroen Verbruggen, Kim-Anh Lê Cao.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>